### ISSUE TITLE
MPboost fails on Ubuntu16.04 with spark standalone

### ISSUE THREAD

**Issue by harshagn** [tizfa/sparkboost#3]
Hi,
    I am facing issue with the sparkboost MPboost execution as shown below.  pom_xml.txt is the pom.xml. Please review it and let me know why i have this issue.  The compilation of sprkboost was successful. 

Please help out with this. 

Config:
Ubuntu16.04, spark-1.6.0 and on Power 8 system. 

[pom_xml.txt](https://github.com/tizfa/sparkboost/files/247471/pom_xml.txt)

Error seen:
hduser@cit923:~/sparkboost$ java -cp ./target/sparkboost-1.0-bundle.jar:./target/sparkboost-1.0*.jar it.tizianofagni.sparkboost.MPBoostLearnerExe  ~/sparkboost/rcv1_topics_train.svm ~/sparkboost/model_output 500 spark://9.37.251.65:7077 24
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
16/05/03 13:57:22 INFO Remoting: Starting remoting
16/05/03 13:57:22 INFO Remoting: Remoting started; listening on addresses :[akka.tcp://sparkDriverActorSystem@9.37.251.65:40192]
[Stage 0:>                                                          (0 + 2) / 2]Exception in thread "main" org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 4 times, most recent failure: Lost task 0.3 in stage 0.0 (TID 6, cit923.rtp.raleigh.ibm.com): java.lang.ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field org.apache.spark.api.java.JavaPairRDD$$anonfun$toScalaFunction$1.fun$1 of type org.apache.spark.api.java.function.Function in instance of org.apache.spark.api.java.JavaPairRDD$$anonfun$toScalaFunction$1
    at java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStreamClass.java:2133)
    at java.io.ObjectStreamClass.setObjFieldValues(ObjectStreamClass.java:1305)
    at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2006)
    at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1924)
    at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
    at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
    at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1924)
    at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
    at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
    at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1924)
    at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
    at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
    at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1924)
    at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
    at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
    at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
    at org.apache.spark.serializer.JavaDeserializationStream.readObject(JavaSerializer.scala:76)
    at org.apache.spark.serializer.JavaSerializerInstance.deserialize(JavaSerializer.scala:115)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:61)
    at org.apache.spark.scheduler.Task.run(Task.scala:89)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

Driver stacktrace:
    at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1431)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1419)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1418)
    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
    at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1418)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
    at scala.Option.foreach(Option.scala:236)
    at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:799)
    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1640)
    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1599)
    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1588)
    at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
    at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:620)
    at org.apache.spark.SparkContext.runJob(SparkContext.scala:1832)
    at org.apache.spark.SparkContext.runJob(SparkContext.scala:1952)
    at org.apache.spark.rdd.RDD$$anonfun$reduce$1.apply(RDD.scala:1025)
    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:111)
    at org.apache.spark.rdd.RDD.withScope(RDD.scala:316)
    at org.apache.spark.rdd.RDD.reduce(RDD.scala:1007)
    at org.apache.spark.api.java.JavaRDDLike$class.reduce(JavaRDDLike.scala:370)
    at org.apache.spark.api.java.AbstractJavaRDDLike.reduce(JavaRDDLike.scala:46)
    at it.tizianofagni.sparkboost.DataUtils.computeNumFeatures(DataUtils.java

Thanks
Harsha

**Comment by harshagn** [comment]
resolution is as follows. I could resolve this with help of this on-line help.
http://stackoverflow.com/questions/28186607/java-lang-classcastexception-using-lambda-expressions-in-spark-job-on-remote-ser

Hence made below code change while executing the MPboostClassifier during building the model. I think similar changes in AdaBoostClassifier build will also help.

SparkConf conf = new SparkConf().setAppName("Spark MPBoost learner").set("spark.eventLog.enabled", "true").setJars(new String[]{"/home/hduser/sparkboost/target/sparkboost-1.0-bundle.jar"}).setMaster("spark://9.37.251.65:7077");
        //conf.setMaster(sparkMaster); //commented as already the previous string has spark url

Thanks
Harsha

**Comment by tizfa** [comment]
Thanks, I had this problem too in the past, I fixed it in the same way but in the meantime, in the develop branch, I changed a lot of things, in particular now the main applications are only launched using spark-submit. Spark-submit automatically copy the jar package on overy host of the cluster where the program  should run, so this problem is no more present. I think I will update the release branch with the new code in the next few days (unfortunately currently I am very busy with other activities).
